builtin/v13/migration/market.go

```diff
@@ -0,0 +1,409 @@
+				return cid.Undef, cid.Undef, xerrors.Errorf("failed to unmarshal old state: %w", err)
+			}
+			ok, err := prevOutStates.Get(uint64(deal), &prevNewState)
+			if err != nil {
+				return cid.Undef, cid.Undef, xerrors.Errorf("failed to get previous newstate: %w", err)
+			}
+			if !ok {
+				return cid.Undef, cid.Undef, xerrors.Errorf("failed to get previous newstate: not found")
+			}
+
+			newState := market13.DealState{
+				SectorNumber:     prevNewState.SectorNumber,
+				SectorStartEpoch: oldState.SectorStartEpoch,
+				LastUpdatedEpoch: oldState.LastUpdatedEpoch,
+				SlashEpoch:       oldState.LastUpdatedEpoch,
+			}
+
+			// if the deal is now slashed, UNSET the sector number
+			if newState.SlashEpoch != -1 {
+				newState.SectorNumber = 0
+			}
+
+			if err := prevOutStates.Set(uint64(deal), &newState); err != nil {
+				return cid.Undef, cid.Undef, xerrors.Errorf("failed to set new state: %w", err)
+			}
+		}
+	}
+
+	// now, process updatesToMinerToSectorToDeals, making the appropriate updates
+
+	for miner, sectorUpdates := range m.providerSectors.updatesToMinerToSectorToDeals {
+		var actorSectorsMapRoot typegen.CborCid
+		found, err := prevOutProviderSectors.Get(abi.UIntKey(uint64(miner)), &actorSectorsMapRoot)
+		if err != nil {
+			return cid.Undef, cid.Undef, xerrors.Errorf("failed to get actor sectors: %w", err)
+		}
+
+		var actorSectors *adt13.Map
+		if !found {
+			// can happen, this miner just didn't have any sectors before
+			actorSectors, err = adt13.MakeEmptyMap(ctxStore, market13.ProviderSectorsHamtBitwidth)
+			if err != nil {
+				return cid.Undef, cid.Undef, xerrors.Errorf("failed to make empty actor sectors map: %w", err)
+			}
+		} else {
+			actorSectors, err = adt13.AsMap(ctxStore, cid.Cid(actorSectorsMapRoot), market13.ProviderSectorsHamtBitwidth)
+			if err != nil {
+				return cid.Undef, cid.Undef, xerrors.Errorf("failed to load actor sectors map: %w", err)
+			}
+		}
+
+		var sectorDealIDs market13.SectorDealIDs
+		for sector, deals := range sectorUpdates {
+			// This signals the unlikely case where a sector with deals got GCed after the premigration, delete the entry for it
+			if deals == nil {
+				err = actorSectors.Delete(miner13.SectorKey(sector))
+				if err != nil {
+					return cid.Undef, cid.Undef, xerrors.Errorf("failed to delete actorSectors entry: %w", err)
+				}
+
+				continue
+			} else {
+				sectorDealIDs = deals
+				if err := actorSectors.Put(miner13.SectorKey(sector), &sectorDealIDs); err != nil {
+					return cid.Undef, cid.Undef, xerrors.Errorf("failed to put sector: %w", err)
+				}
+			}
+		}
+
+		// check if actorSectors are empty
+		err = actorSectors.ForEach(nil, func(k string) error {
+			return errItemFound
+		})
+		nonEmpty := false
+		if err == errItemFound {
+			nonEmpty = true
+			err = nil
+		}
+		if err != nil {
+			return cid.Undef, cid.Undef, xerrors.Errorf("failed to iterate actor sectors: %w", err)
+		}
```

Comment:
Consider attaching this to the "map" definition in `util/adt/map.go` so I never have to look at this code again?